File: solutions/csharp/change/1/Change.cs
public static class Change
{
    public static int[] FindFewestCoins(int[] coins, int target)
    {
        EnsureChange(coins, target);

        var best = new List<int>?[target + 1];
        best[0] = [];

        foreach (var amount in Enumerable.Range(1, target))
        {
            best[amount] = coins
                .Where(coin => coin <= amount)
                .Select(coin => AddCoin(best[amount - coin], coin))
                .OfType<List<int>>()
                .MinBy(candidate => candidate.Count);
        }

        return best[target]?.Order().ToArray() ?? throw new ArgumentException();
    }

    private static void EnsureChange(int[] coins, int target)
    {
        if (target < 0 || target > 0 && coins.All(coin => coin > target))
            throw new ArgumentException();
    }

    private static List<int>? AddCoin(List<int>? coins, int coin)
    {
        if (coins is null)
            return null;

        return [.. coins, coin];
    }
}

File: solutions/scala/change/1/src/main/scala/Change.scala
object Change {
  private type Coins = List[Int]

  def findFewestCoins(target: Int, coins: Coins): Option[Coins] =
    if target < 0 then None
    else
      val sortedCoins = coins.sorted
      val best = Array.fill[Option[Coins]](target + 1)(None)
      best(0) = Some(List.empty)

      for amount <- 1 to target do
        best(amount) = bestChangeFor(amount, sortedCoins, best)

      best(target)

  private def bestChangeFor(amount: Int, coins: Coins, best: Array[Option[Coins]]): Option[Coins] =
    coins
      .filter(_ <= amount)
      .flatMap(coin => best(amount - coin).map(change => coin +: change))
      .sortBy(_.length)
      .headOption
}
